### PR TITLE
fix: properly truncate timestamps to beginning of window bounds

### PR DESCRIPTION
**Pull request by jlapacik**
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: execute/window.go [hunk 1]
```diff
@@ -106,7 +106,11 @@ func (w Window) GetOverlappingBounds(b Bounds) []Bounds {
...
 // of nanoseconds.
 func truncateByNsecs(t Time, d Duration) Time {
-	remainder := int64(t) % d.Nanoseconds()
+	dur := d.Nanoseconds()
+	remainder := int64(t) % dur
+	if remainder < 0 {
+		remainder += dur
+	}
 	return t - Time(remainder)
 }
```
Comment (jlapacik) [inline execute/window.go:115]: `%` can return a negative value so we need to normalize the remainder as `t - Time(remainder)` expects a non-negative input.

File: execute/window_test.go [hunk 2]
```diff
@@ -208,6 +208,54 @@ func TestWindow_GetEarliestBounds(t *testing.T) {
 				Stop:  execute.Time(5*time.Minute + 30*time.Second),
 			},
 		},
+		{
+			name: "truncate before offset",
+			w: MustWindow(
+				values.ConvertDurationNsecs(5*time.Second),
+				values.ConvertDurationNsecs(5*time.Second),
+				values.ConvertDurationNsecs(2*time.Second), false),
```
Comment (nathanielc) [inline execute/window_test.go:216]: Was refreshing my memory on what MustWindow does and noticed that the bool arg months is not used. Should we remove that as part of this change?

File: values/time.go [hunk 3]
```diff
@@ -212,6 +212,14 @@ func (d Duration) Nanoseconds() int64 { return d.nsecs }
 // It will ensure that the output duration is the smallest positive
 // duration that is the equivalent of the current duration.
 func (d Duration) Normalize(interval Duration) Duration {
+	if d.Months() != 0 && d.Nanoseconds() == 0 &&
+		interval.Months() != 0 && interval.Nanoseconds() == 0 {
+		return Duration{
+			months:   d.Months() % interval.Months(),
```
Comment (nathanielc) [inline values/time.go:218]: Below the nanosecond logic checks for negative offsets. Do we need to do the same here for months?

File: execute/window_test.go [hunk 4]
```diff
@@ -208,6 +208,54 @@ func TestWindow_GetEarliestBounds(t *testing.T) {
...
+			w: MustWindow(
+				values.ConvertDurationMonths(5),
+				values.ConvertDurationMonths(5),
+				values.ConvertDurationMonths(2), true),
+			t: mustParseTime("1970-04-01T00:00:00Z"),
+			want: execute.Bounds{
+				Start: mustParseTime("1970-03-01T00:00:00Z"),
+				Stop:  mustParseTime("1970-08-01T00:00:00Z"),
+			},
+		},
```
Comment (nathanielc) [inline execute/window_test.go:258]: Could you add a few more test cases where the offset itself is negative? There seem to be specific code branches for the case where offset is negative and I want to make sure we are testing that code path.

**Review by nathanielc — CHANGES_REQUESTED** [review]
Looks great and the change makes sense. I have requested a few more tests cases to ensure we are covering more edge cases.

File: execute/window_test.go (hunk 2 above)
Comment (jlapacik) [reply execute/window_test.go:216]: I did not notice that, yes I'll remove it. I wish the go compiler would identify unused function params.

File: values/time.go (hunk 3 above)
Comment (jlapacik) [reply values/time.go]: I added a check for negative offsets.

File: execute/window_test.go (hunk 4 above)
Comment (jlapacik) [reply execute/window_test.go:258]: We already had one for nanoseconds, I added one for calendar durations.

**Review by nathanielc — APPROVED** [review]
